x-pack/plugins/maps/public/connected_components/timeslider/timeslider.tsx

```diff
@@ -59,28 +52,25 @@ export class Timeslider extends Component<Props, {}> {
     }
 
     this._subscriptions.add(
-      controlGroup
-        .getOutput$()
+      controlGroup.timeslice$
         .pipe(
-          distinctUntilChanged(({ timeslice: timesliceA }, { timeslice: timesliceB }) =>
-            _.isEqual(timesliceA, timesliceB)
-          )
+          tap(() => {
+            this.dataLoading = true;
+          }),
+          switchMap((timeslice) => {
+            this.props.setTimeslice(
+              timeslice === undefined
+                ? undefined
+                : {
+                    from: timeslice[0],
+                    to: timeslice[1],
+                  }
+            );
+            return this.props.waitForTimesliceToLoad$;
+          })
         )
-        .subscribe(({ timeslice }) => {
-          // use waitForTimesliceToLoad$ observable to wait until next frame loaded
-          // .pipe(first()) waits until the first value is emitted from an observable and then automatically unsubscribes
-          this.props.waitForTimesliceToLoad$.pipe(first()).subscribe(() => {
-            controlGroup.anyControlOutputConsumerLoading$.next(false);
-          });
-
-          this.props.setTimeslice(
-            timeslice === undefined
-              ? undefined
-              : {
-                  from: timeslice[0],
-                  to: timeslice[1],
-                }
-          );
+        .subscribe(() => {
+          this.dataLoading = false;
```

Comment:
dataLoading needs to be part of state to trigger a re-render and pass updated value to ControlGroupRenderer. If you want, we can just re-write this file with hooks. Below is a version I have working.

```
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

import React, { useEffect, useState } from 'react';
import { Observable, switchMap, tap } from 'rxjs';
import {
  type ControlGroupStateBuilder,
  ControlGroupRenderer,
  type ControlGroupRuntimeState,
  ControlGroupRendererApi,
} from '@kbn/controls-plugin/public';
import type { TimeRange } from '@kbn/es-query';
import { Timeslice } from '../../../common/descriptor_types';

export interface Props {
  setTimeslice: (timeslice?: Timeslice) => void;
  timeRange: TimeRange;
  waitForTimesliceToLoad$: Observable<void>;
}

export function Timeslider({ setTimeslice, timeRange, waitForTimesliceToLoad$ }: Props) {
  const [dataLoading, setDataLoading] = useState(false);
  const [api, setApi] = useState<ControlGroupRendererApi | undefined>();

  useEffect(() => {
    if (!api) {
      return;
    }
    let canceled = false;

    const subscription = api.timeslice$
      .pipe(
        tap(() => {
          setDataLoading(true);
        }),
        switchMap((timeslice) => {
          setTimeslice(
            timeslice === undefined
              ? undefined
              : {
                  from: timeslice[0],
                  to: timeslice[1],
                }
          );
          return waitForTimesliceToLoad$;
        })
      )
      .subscribe(() => {
        setDataLoading(false);
      });

    return () => {
      subscription?.unsubscribe();
      canceled = true;
    };
  }, [api, setTimeslice, waitForTimesliceToLoad$]);

  return (
    <div className="mapTimeslider mapTimeslider--animation">
      <ControlGroupRenderer
        onApiAvailable={(nextApi: ControlGroupRendererApi) => {
          setApi(nextApi);
        }}
        dataLoading={dataLoading}
        getCreationOptions={async (
          initialState: Partial<ControlGroupRuntimeState>,
          builder: ControlGroupStateBuilder
        ) => {
          builder.addTimeSliderControl(initialState);
          return {
            initialState,
          };
        }}
        timeRange={timeRange}
      />
    </div>
  );
}

```

Reply:
I'm confused - it **is** part of state, isn't it? I tested this, and it seems to be working as expected - it waits for `dataLoading` to be `false` before outputting the next time slice 🤔 I'd be fine switching to hooks - just curious what the problem is with the current implementation

Reply:
No `this.dataLoading` is not part of state. This is a class based component and state is accessed via `this.state.`